### PR TITLE
More of Player

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/object/update/autodepositupdate.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    virtual NameKeyType Get_Module_Name_Key() const override;
+
+    virtual void CRC_Snapshot(Xfer *xfer) override;
+    virtual void Xfer_Snapshot(Xfer *xfer) override;
+    virtual void Load_Post_Process() override;
+
+    virtual UpdateSleepTime Update() override;
+
+    void Award_Initial_Capture_Bonus(Player *p);
+};
```
Comment (xezon) [inline src/game/logic/object/update/autodepositupdate.h]: eol

File: src/game/logic/object/object.cpp [hunk 2]
```diff
@@ -571,3 +571,83 @@ void Object::Update_Upgrade_Modules()
...
+{
+    unsigned int condition = m_weaponBonusCondition;
+    m_weaponBonusCondition = ~(1 << bonus) & condition;
+
+    if (condition != m_weaponBonusCondition) {
+        m_weaponSet.Weapon_Set_On_Weapon_Bonus_Change(this);
+    }
+}
+
+float Object::Get_Shroud_Clearing_Range() const
```
Comment (xezon) [inline src/game/logic/object/object.cpp:625]: `Get_Shroud_Clearing_Range` is sorted above `Set_Shroud_Clearing_Range` function in header.

File: src/game/logic/ai/aiplayer.h [hunk 3]
```diff
@@ -17,38 +17,203 @@
...
+        DLINK_TeamReadyQueue() : m_prev(nullptr), m_next(nullptr) {}
+        ~DLINK_TeamReadyQueue()
+        {
+            captainslog_dbgassert(m_prev == nullptr && m_next == nullptr, "destroying dlink still in a list TeamReadyQueue");
+        }
+    };
+
+    DLINK_TeamBuildQueue m_dlink_TeamBuildQueue;
+    DLINK_TeamReadyQueue m_dlink_TeamReadyQueue;
+    WorkOrder *m_workorders;
```
Comment (xezon) [inline src/game/logic/ai/aiplayer.h:183]: m_workOrders?

File: src/game/logic/ai/aiplayer.cpp [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+    }
+}
+
+AIPlayer::AIPlayer(Player *p) :
+    m_player(p),
+    m_readyToBuildTeam(false),
+    m_readyToBuildStructure(false),
+    m_teamTimer(2),
+    m_structureTimer(2),
+    m_teamDelaySeconds(10),
```
Comment (xezon) [inline src/game/logic/ai/aiplayer.cpp:37]: This correct? It gets overwritten in constructor body.

File: src/game/logic/ai/aiplayer.cpp [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+    m_buildDelay(0),
+    m_teamDelay(0),
+    m_skillsetSelector(INVALID_SKILLSET_SELECTION),
+    m_dozerID(OBJECT_UNK),
```
Comment (xezon) [inline src/game/logic/ai/aiplayer.cpp:41]: m_baseRadius not initialized

File: src/game/logic/ai/aiplayer.cpp [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+    m_structureRepairCount(0),
+    m_dozerQueued(false),
+    m_supplySourceFrame(0),
+    m_supplySourceID(OBJECT_UNK),
+    m_supplyCenterID(OBJECT_UNK)
+{
+    m_frameLastBuildingBuilt = g_theGameLogic->Get_Frame();
+    p->Set_Can_Build_Units(false);
+
+    for (int i = 0; i < 2; i++) {
```
Comment (xezon) [inline src/game/logic/ai/aiplayer.cpp:54]: ARRAY_SIZE(m_structureRepairQueue)

File: src/game/logic/ai/aiplayer.cpp [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+        m_structureRepairQueue[i] = OBJECT_UNK;
+    }
+
+    m_dozerPosition.Zero();
```
Comment (xezon) [inline src/game/logic/ai/aiplayer.cpp:58]: m_dozerID not initialized.

File: src/game/logic/ai/aiplayer.cpp [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+    m_teamTimer(2),
+    m_structureTimer(2),
+    m_teamDelaySeconds(10),
+    m_buildDelay(0),
+    m_teamDelay(0),
+    m_skillsetSelector(INVALID_SKILLSET_SELECTION),
+    m_dozerID(OBJECT_UNK),
+    m_structureRepairCount(0),
+    m_dozerQueued(false),
+    m_supplySourceFrame(0),
```
Comment (xezon) [inline src/game/logic/ai/aiplayer.cpp:44]: m_dozerRepairing
m_bridgeRepairTimer
m_supplySourceFrame

not initialized

File: src/game/common/rts/player.cpp [hunk 9]
```diff
@@ -1937,3 +2053,221 @@ ScienceAvailabilityType Player::Get_Science_Availability_Type_From_String(const
...
+                }
+            }
+        }
+    }
+}
+
+void Player::Remove_Battle_Plan_Bonuses_For_Object(Object *obj)
+{
+    BattlePlanBonuses *bonuses = new BattlePlanBonuses();
+    bonuses = m_battlePlanBonuses;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: Something is not right here. `bonuses` will leak straight away.

File: src/game/common/rts/player.cpp [hunk 10]
```diff
@@ -1937,3 +2053,221 @@ ScienceAvailabilityType Player::Get_Science_Availability_Type_From_String(const
...
+    info->Set_Template_Name(template_name);
+    info->Set_Location(*loc);
+    info->Set_Angle(angle);
+    info->Set_Unk4();
+    info->Set_Num_Rebuilds(1);
+    info->Set_Next(m_buildListInfo);
+    m_buildListInfo = info;
+}
+
+void Player::Becoming_Team_Member(Object *thg, bool yes)
```
Comment (xezon) [inline src/game/common/rts/player.cpp:2081]: Object *obj,
bool add

File: src/game/common/rts/player.cpp [hunk 11]
```diff
@@ -1937,3 +2053,221 @@ ScienceAvailabilityType Player::Get_Science_Availability_Type_From_String(const
...
+{
+    if (thg != nullptr) {
+        if (!thg->Get_Status_Bits().Test(OBJECT_STATUS_UNDER_CONSTRUCTION)) {
+            thg->Friend_Adjust_Power_For_Player(yes);
+        }
+
+        if (this != g_thePlayerList->Get_Neutral_Player()) {
+            if (yes) {
+                AutoDepositUpdate *module =
+                    static_cast<AutoDepositUpdate *>(thg->Find_Module(Name_To_Key("AutoDepositUpdate")));
```
Comment (xezon) [inline src/game/common/rts/player.cpp:2091]: key var is not held in static?

File: src/game/common/rts/player.cpp [hunk 12]
```diff
@@ -1937,3 +2053,221 @@ ScienceAvailabilityType Player::Get_Science_Availability_Type_From_String(const
...
+    m_buildListInfo = info;
+}
+
+void Player::Add_To_Priority_Build_List(Utf8String template_name, Coord3D *loc, float angle)
+{
+    BuildListInfo *info = new BuildListInfo();
+    info->Set_Template_Name(template_name);
+    info->Set_Location(*loc);
+    info->Set_Angle(angle);
+    info->Set_Unk4();
```
Comment (xezon) [inline src/game/common/rts/player.cpp:2075]: `Set_Priority` perhaps.

File: src/game/common/rts/player.cpp [hunk 13]
```diff
@@ -717,12 +722,29 @@ Team *Player::Get_Default_Team()
...
+        if (change) {
+            obj->Set_Disabled(DISABLED_TYPE_DISABLED_UNDERPOWERED);
+        } else {
+            obj->Clear_Disabled(DISABLED_TYPE_DISABLED_UNDERPOWERED);
+        }
+    }
+}
+
 // zh: 0x00456820, wb: 0x0086048D
 void Player::On_Power_Brown_Out_Change(bool change)
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: bool power_disable

File: src/game/common/rts/player.cpp [hunk 14]
```diff
@@ -717,12 +722,29 @@ Team *Player::Get_Default_Team()
     return m_defaultTeam;
 }
 
+void Do_Power_Disable(Object *obj, void *data)
+{
+    bool change = *static_cast<bool *>(data);
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: bool power_disable

File: src/game/logic/object/update/autodepositupdate.h (hunk 1 above)
Comment (jonwil) [reply src/game/logic/object/update/autodepositupdate.h]: fixed.

File: src/game/logic/ai/aiplayer.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/logic/ai/aiplayer.cpp:37]: That's correct.

File: src/game/logic/ai/aiplayer.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/logic/ai/aiplayer.cpp:41]: fixed.

File: src/game/logic/ai/aiplayer.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/ai/aiplayer.cpp:44]: fixed.

File: src/game/common/rts/player.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:2091]: Nope.

File: src/game/common/rts/player.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

**Review by xezon — APPROVED** [review]
Tested Radar, Power Outage, Strategy Center Bonuses, Skirmish AI enemy, building base. No new obvious issues encountered. No crashes. Tested in RelWithDebInfo.

Looks good to me.

**Pull request merged**